### PR TITLE
Feature/slider

### PR DESCRIPTION
**Pull request by ydomenjoud**
create a real slider
do not forget to import hammerjs in main.ts of your application to enable gesture support

### CODE REVIEW

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts [hunk 1]
```diff
@@ -22,192 +21,183 @@ import { buildProviders, FormControlAccessComponent } from '../../utils/index';
...
 
-  private tickIntervalValue = 0;
+  @ViewChild('sliderWrapper') private sliderWrapper: ElementRef;
 
-  private valueChangeSubscription = Subscription.EMPTY;
+  private _value = 0;
 
-  constructor(private cdr: ChangeDetectorRef) {
-    super();
+  get value() {
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts:55]: @Input()

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts [hunk 2]
```diff
@@ -22,192 +21,183 @@ import { buildProviders, FormControlAccessComponent } from '../../utils/index';
...
+    this.value = this.positionInSliderToValue(position);
   }
 
-  /**
-   * @inheritDoc
-   */
-  setDisabledState(isDisabled: boolean): void {
-    this.isDisabled = isDisabled;
+  selectionContainerStyles(): { 'width.%': number, backgroundColor?: string } {
+    const styles: { 'width.%': number, backgroundColor?: string } = {
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts:131]: const styles: { 'width.%': number, backgroundColor?: string } = {
     'width.%': this.valueToPercent(this.value),
     backgroundColor = this.color || undefined
}

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts [hunk 3]
```diff
@@ -22,192 +21,183 @@ import { buildProviders, FormControlAccessComponent } from '../../utils/index';
...
-   */
-  getLimitValue(value: number): number {
-    return this.limit !== 0 ? Math.min(this.limit, value) : value;
+   * convert slide position x to percent of the slider container width
+   * if slider is 400px width, and @input to = 1200,
+   * a slide event with x=200px should be a value of 600
+   * a slide event with x=300px should be a value of 900
+   * a slide event with x=400px should be a value of 1200
+   **/
+  private positionInSliderToValue(position) {
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts]: type parameters

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts [hunk 4]
```diff
@@ -22,192 +21,183 @@ import { buildProviders, FormControlAccessComponent } from '../../utils/index';
...
+    const percent = this.clamp(position, 0, this.sliderDimensions.width) / this.sliderDimensions.width;
+    return percent * this.to;
+  }
+
+  private getClosestValue(current) {
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts]: type parameters

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/README.md [hunk 5]
```diff
@@ -2,10 +2,19 @@
...
 
-# Specifications
 
-* Should have a min and a max
-* Should have a default value
-* May display thumb label
-* May show ticks
-* May block to a limit value
+# Gesture Support
+kal-slider rely on HammerJS for gestures.
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/README.md]: relies

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts [hunk 6]
```diff
@@ -22,192 +21,183 @@ import { buildProviders, FormControlAccessComponent } from '../../utils/index';
...
   }
 
-  ngOnInit() {
-    this.valueChangeSubscription = this.control.valueChanges.subscribe(
-      value => {
-        if (this.disabled) {
-          return;
-        }
+  private increment(numSteps: number) {
+    this.value = this.value + numSteps * this.tick;
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts]: this.value += numSteps * this.tick;

File: src/main.ts [hunk 7]
```diff
@@ -4,6 +4,8 @@ import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 import { AppModule } from './app/app.module';
 import { environment } from './environments/environment';
 
+import 'hammerjs'; // required for gesture
```
Comment (MichaelGrosio) [inline src/main.ts]: imports

**Review by MichaelGrosio — CHANGES_REQUESTED** [review]

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts [hunk 8]
```diff
@@ -22,192 +21,183 @@ import { buildProviders, FormControlAccessComponent } from '../../utils/index';
...
-  getLimitValue(value: number): number {
-    return this.limit !== 0 ? Math.min(this.limit, value) : value;
+   * convert slide position x to percent of the slider container width
+   * if slider is 400px width, and @input to = 1200,
+   * a slide event with x=200px should be a value of 600
+   * a slide event with x=300px should be a value of 900
+   * a slide event with x=400px should be a value of 1200
+   **/
+  private positionInSliderToValue(position) {
+    const percent = this.clamp(position, 0, this.sliderDimensions.width) / this.sliderDimensions.width;
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts]: we have a clamp function in utils

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts [hunk 9]
```diff
@@ -22,192 +21,183 @@ import { buildProviders, FormControlAccessComponent } from '../../utils/index';
...
-      }
-    );
+  private valueToPercent(value: number) {
+    return (value / this.to) * 100;
   }
 
-  ngOnDestroy(): void {
-    this.valueChangeSubscription.unsubscribe();
+  /** Return a number between two numbers. */
+  private clamp(value: number, min = 0, max = 1) {
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts]: remove this function

**Review by jsp-up — CHANGES_REQUESTED** [review]

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts (hunk 2 above)
Comment (ydomenjoud) [reply projects/kalidea/kaligraphi/src/lib/atoms/kal-slider/kal-slider.component.ts:131]: no, it will erase style by css

**Review by Bellico — DISMISSED** [review]

**Review by jsp-up — APPROVED** [review]

**Review by MichaelGrosio — APPROVED** [review]

**Pull request merged**